Fix release models compile, and test failure from dependency regression

kb2ma opened the pull request:
See issue below for details on compilation failure. Also includes a cosmetic fix from prettify.

This fix minimally narrows the `body` input parameters by keeping them generics. It would be simpler to directly describe the body parameters as AnyObject without the generics, but I don't have enough context to know if that solution is better.

The test failure is due to a regression in dockerode v3.3.4, as described in [this issue](https://github.com/apocas/dockerode/issues/696). We have proposed a [fix PR](https://github.com/apocas/dockerode/issues/696). While waiting for a response, this PR pins dockerode to v3.3.3.

Fixes #10 , #15

kb2ma committed:
Fix underspecified generics in release/models

Change-type: patch
Signed-off-by: Ken Bannister <kb2ma@runbox.com>

dfunckt reviewed (commented):
The fix seems reasonable to me, the bug I'm curious why it only now manifested itself -- this code has been like that for ages.

Feel free to merge at will, regardless.

kb2ma commented:
Thanks for the review, @dfunckt . As you can see below the CI tests are failing, and I can recreate locally. I see a lot of timeout failures. Below is the [first example](https://github.com/balena-io-modules/balena-compose/blob/159cdbad4f9c91edd1b9b537a62cdfa1e3dcf309/test/build/all.spec.ts#L95). I'm new to this code. Does anything jump out at you as the cause?

```
  Directory build
    1) should build a directory image


  0 passing (1m)
  1 failing

  1) Directory build
       should build a directory image:
     Error: Timeout of 60000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves. (/home/kbee/dev/balena-compose/repo/test/build/all.spec.ts)
      at listOnTimeout (internal/timers.js:557:17)
      at processTimers (internal/timers.js:500:7)
```

kb2ma commented:
@dfunckt , with a pointer from @pipex, I've found the compilation problem is due to a stricter check in TypeScript 4.8 -- [Unconstrained Generics No Longer Assignable to {}](https://devblogs.microsoft.com/typescript/announcing-typescript-4-8/#unconstrained-generics-no-longer-assignable-to). In this case the `body` param is a Params object in which all members are optional. So, the new constraint ensures the passed in object is not null.

Still need to fix the test failures.

kb2ma commented:
Tests fail with dependency [dockerode](https://github.com/apocas/dockerode) v3.3.4; succeed with v3.3.3. Now need to understand why.

kb2ma commented:
@pipex, @dfunckt : I understand more about the test failure, but not enough to fix it. The problem is with the code below, from dockerode's [docker.js](https://github.com/apocas/dockerode/blob/ad170f5d659744fd19ef6e6bab64868480a6e04d/lib/docker.js#L285), new in v3.3.4. I have added a couple of `console.log()` statements, and show their output below when run from a failing balena-compose [test](https://github.com/balena-io-modules/balena-compose/blob/159cdbad4f9c91edd1b9b537a62cdfa1e3dcf309/test/build/all.spec.ts#L95).

```
 if (callback === undefined) {
    const prepareCtxPromise = new self.modem.Promise(function(resolve, _) {
      util.prepareBuildContext(file, resolve);
      console.log(`file: ${JSON.stringify(file)}`);
    });

    return prepareCtxPromise.then((ctx)=> {
      console.log(`ctx: ${JSON.stringify(ctx)}`);
      return new self.modem.Promise(function(resolve, reject) {
        optsf.file = ctx;
        self.modem.dial(optsf, function(err, data) {
          if (err) {
            return reject(err);
          }
          resolve(data);
        });
      });
    })
  } else {
    util.prepareBuildContext(file, (ctx) => {
      optsf.file = ctx;
      self.modem.dial(optsf, function(err, data) {
        callback(err, data);
      });
    })
  }
```

Log output below. Notice in the `ctx:` line that the object has `false` for the `readable` and `writeable` properties. So, something happens after resolving `prepareCtxPromise`.

```
file: {"_events":{"end":[null,null,null],"close":[null,null],"error":[null,null]},"_eventsCount":5,"writable":true,"readable":true,"paused":false,"autoDestroy":true}

ctx: {"_events":{"end":[null,null,null],"close":[null,null],"error":[null,null]},"_eventsCount":5,"writable":false,"readable":false,"paused":false,"autoDestroy":true}
```

Strictly speaking, the creation of `prepareCtxPromise` is not essential. [Here](https://github.com/kb2ma/dockerode/blob/53ff9158772af000798e18f91638db1df98c2a57/lib/docker.js#L285) is a modified version without the separate `prepareCtxPromise`, and this works fine from the balena-compose tests.

However, I think the problem is that balena-compose is using the file/ctx object in a way that dockerode's `buildImage()` function does not expect. balena-compose passes a Duplex type via event-stream, as you can see [here](https://github.com/balena-io-modules/balena-compose/blob/159cdbad4f9c91edd1b9b537a62cdfa1e3dcf309/lib/build/builder.ts#L103) to mimic (?) the [expected](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/241890973f0068ebd073027641ac778899cdea4a/types/dockerode/index.d.ts#L1810) ReadableStream. We don't [push data](https://github.com/balena-io-modules/balena-compose/blob/159cdbad4f9c91edd1b9b537a62cdfa1e3dcf309/lib/build/builder.ts#L193) into the stream until later.

So, at the moment the problem is with the way that balena-compose uses dockerode. Is dockerode being too strict in what they expect from a ReadableStream?

dfunckt commented:
@kb2ma `DuplexStream` *is* a `ReadableStream` and it should be accepted. Without digging deep into the details, it appears to be a regression in Dockerode introduced by that PR.

pipex commented:
The latest bump to dockerode also bumped docker-modem, with this fix https://github.com/apocas/docker-modem/commit/b3f33cd565a1b6014db44dcd55108fb1bf53352a for the following issue https://github.com/apocas/docker-modem/issues/142. I wonder if balena-compose was relying on the callback being triggered multiple times.

One test you might want to do @kb2ma is to downgrade just docker-modem under `node_modules/dockerode` and test again.

dfunckt commented:
> I wonder if balena-compose was relying on the callback being triggered multiple times.

I can't think of any reason that would make invoking the callback more than once desirable, so I don't think this is the case.

kb2ma commented:
Thanks for the suggestion to downgrade docker-modem @pipex. I agree that the changes in v3.0.6 look like they might be relevant. However, I tested with docker-modem v3.0.3 through v3.0.5, and none of them worked either.

My approach at this point is that there is an issue with the extra layer of promise in the `prepareCtxPromise` variable shown above that does not support the balena-compose use of `buildImage()` with a Duplex as a ReadableStream. I will try to create a simple test to recreate the failure for the dockerode repo.

kb2ma commented:
Creating simple example in balena-io-playground/dockerode-build-image-hang [repo](https://github.com/balena-io-playground/dockerode-build-image-hang).

kb2ma commented:
Created an [issue](https://github.com/apocas/dockerode/pull/696) / [fix](https://github.com/apocas/dockerode/issues/695) for Dockerode to resolve test failures.

Page- commented on lib/release/models.ts:
```diff
@@ -114,15 +119,15 @@ export function getOrCreate<T, U, V extends Filter>(
 }
 
-export function update<T>(
+export function update<T extends AnyObject>(
```
```suggestion
export function update<T extends {}>(
```

Page- commented on lib/release/models.ts:
```diff
@@ -114,15 +119,15 @@ export function getOrCreate<T, U, V extends Filter>(
 }
 
-export function create<T, U>(
+export function create<T, U extends AnyObject>(
```
```suggestion
export function create<T, U extends {}>(
```

Page- commented on lib/release/models.ts:
```diff
@@ -95,7 +100,7 @@ export interface ReleaseImageModel extends ReleaseImageAttributesBase {
 // Helpers
 
-export function getOrCreate<T, U, V extends Filter>(
+export function getOrCreate<T, U extends AnyObject, V extends Filter>(
```
```suggestion
export function getOrCreate<T, U extends {}, V extends Filter>(
```

Page- reviewed (commented):
I'd prefer to use the slightly more relaxed `extends` version which also matches what typescript recommends: https://devblogs.microsoft.com/typescript/announcing-typescript-4-8/#unconstrained-generics-no-longer-assignable-to

kb2ma commented:
> I'd prefer to use the slightly more relaxed extends

Will do. Thanks for the context check.

kb2ma committed:
Prettify fixup

Change-type: patch
Signed-off-by: Ken Bannister <kb2ma@runbox.com>

kb2ma commented:
Squashed changes

kb2ma committed:
Pin dockerode to v3.3.3 to avoid regression

The problem is rework to buildImage() in v3.3.4 that hangs balena-compose use of it.
See Issue: https://github.com/apocas/dockerode/issues/696

Change-type: patch
Signed-off-by: Ken Bannister <kb2ma@runbox.com>
